corehq/apps/hqcase/api.py

```diff
@@ -0,0 +1,210 @@
+    def get_case_id(self):
+        return str(uuid.uuid4()) if self._is_case_creation else self.case_id
+
+    def get_caseblock(self):
+
+        def _if_specified(value):
+            return value if value is not None else CaseBlock.undefined
+
+        return CaseBlock(
+            case_id=self.get_case_id(),
+            user_id=self.user_id,
+            case_type=self.case_type if self._is_case_creation else CaseBlock.undefined,
+            case_name=_if_specified(self.case_name),
+            external_id=_if_specified(self.external_id),
+            owner_id=_if_specified(self.owner_id),
+            create=self._is_case_creation,
+            update=dict(self.properties),
+            index={
+                name: IndexAttrs(index.case_type, index.case_id, index.relationship)
+                for name, index in self.indices.items()
+            },
+        ).as_text()
+
+
+class JsonCaseCreation(BaseJsonCaseChange):
+    case_type = jsonobject.StringProperty(name='@case_type', required=True)
+    temporary_id = jsonobject.StringProperty()
+
+    # overriding from subclass to mark these required
+    case_name = jsonobject.StringProperty(required=True)
+    owner_id = jsonobject.StringProperty(name='@owner_id', required=True)
+
+    _is_case_creation = True
+
+
+class JsonCaseUpdate(BaseJsonCaseChange):
+    case_id = jsonobject.StringProperty(required=True)
+    _is_case_creation = False
+
+
+class UserError(Exception):
+    pass
+
+
+def handle_case_update(domain, data, user, device_id, case_id=None):
+    is_bulk = isinstance(data, list)
+    if is_bulk:
+        updates = _get_bulk_updates(domain, data, user)
+    else:
+        updates = [_get_individual_update(domain, data, user, case_id)]
+
+    xform, cases = _submit_case_updates(updates, domain, user, device_id)
+    if is_bulk:
+        return xform, cases
+    else:
+        return xform, cases[0] if cases else None
+
+
+def _get_individual_update(domain, data, user, case_id=None):
+    if case_id is not None and _missing_cases(domain, [case_id]):
+        raise UserError(f"No case found with ID '{case_id}'")
+
+    try:
+        return _get_case_update(data, user.user_id, case_id)
+    except BadValueError as e:
+        raise UserError(str(e))
+
+
+def _get_bulk_updates(domain, all_data, user):
+    if len(all_data) > 100:
+        raise UserError("You cannot submit more than 100 updates in a single request")
+
+    existing_ids = [c['@case_id'] for c in all_data if isinstance(c, dict) and '@case_id' in c]
+    missing = _missing_cases(domain, existing_ids)
+    if missing:
+        raise UserError(f"The following case IDs were not found: {', '.join(missing)}")
+
+    updates = []
+    errors = []
+    for i, data in enumerate(all_data):
+        try:
+            update = _get_case_update(data, user.user_id, data.pop('@case_id', None))
+            updates.append(update)
+        except BadValueError as e:
+            errors.append(f'Error in row {i}: {e}')
```

Comment:
Is it clear that row numbers are 0-indexed? Otherwise this should be `i+1`.

Reply:
good catch, will update.